File: backend/app/models/project.py
from __future__ import annotations

from datetime import datetime
from enum import Enum

from sqlalchemy import CheckConstraint, DateTime, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class ProjectStatus(str, Enum):
    planning = "planning"
    development = "development"
    staging = "staging"
    production = "production"
    paused = "paused"
    archived = "archived"


class Project(Base):
    __tablename__ = "projects"
    __table_args__ = (
        CheckConstraint(
            "status in ('planning', 'development', 'staging', 'production', 'paused', 'archived')",
            name="ck_projects_status",
        ),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    repo_url: Mapped[str | None] = mapped_column(String(2048), nullable=True)
    production_url: Mapped[str | None] = mapped_column(String(2048), nullable=True)
    status: Mapped[str] = mapped_column(
        String(32),
        nullable=False,
        default=ProjectStatus.planning.value,
        index=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

File: frontend/src/types/project.ts
export const projectStatuses = ["planning", "development", "staging", "production", "paused", "archived"] as const;

export type ProjectStatus = (typeof projectStatuses)[number];

export interface Project {
  id: number;
  name: string;
  description: string | null;
  repo_url: string | null;
  production_url: string | null;
  status: ProjectStatus;
  created_at: string;
  updated_at: string;
}

export interface ProjectInput {
  name: string;
  description?: string | null;
  repo_url?: string | null;
  production_url?: string | null;
  status: Exclude<ProjectStatus, "archived">;
}
